pkg/utils/tasks/tasks.go

```diff
@@ -72,13 +110,64 @@ func (tm *TaskManager) IsStopped() bool {
 }
 
 // Queue a task with the specified context. Blocks if the queue is full.
-func (tm *TaskManager) QueueContext(ctx context.Context, task Task) {
-	tm.queue <- &TaskHandler{task, ctx}
+func (tm *TaskManager) QueueContext(ctx context.Context, task Task, opts ...Option) {
+	conf := options{
+		Backoff: &backoff.ZeroBackOff{},
+	}
+	for _, opt := range opts {
+		opt(&conf)
+	}
+
+	// Wrap the task with retry logic
+	retry := func(ctx context.Context) (err error) {
+		var retries int
+	retryLoop:
+		for {
+			if err = task.Do(ctx); err == nil {
+				return nil
+			}
+
+			retries++
+			if retries >= conf.Retries {
+				break retryLoop
+			}
+
+			// Wait for the backoff duration before retrying
+			// Note: This blocks the worker thread so queue sizes should be large
+			// enough to avoid blocking new tasks from being queued.
+			wait := time.After(conf.Backoff.NextBackOff())
+
+			select {
+			case <-ctx.Done():
+				err = ctx.Err()
+				break retryLoop
+			case <-wait:
+			}
```

Comment:
This will block the worker thread, are we okay with this?

Reply:
Well this is easier said than done...I've been trying to come up with a solution where retry tasks can be added to the queue but still allow the `Stop()` to wait for the workers to exit. What makes it hard is that the channel can't be closed while retry tasks are still being added.

Reply:
happy to talk this through if you'd like